libc/vga/tty-graph.c

```diff
@@ -0,0 +1,287 @@
+    char *cplotter = tty->fb + gy1 * xsfb + gx1 * sizeof(TtyBgrxColor);
+    const char *creader = tty->canvas + gy1 * xs
+                                      + gx1 * sizeof(TtyCanvasColor);
+    TtyResetDirty(tty);
+    while (yleft-- != 0) {
+      TtyBgrxColor *plotter = (TtyBgrxColor *)cplotter;
+      const TtyCanvasColor *reader = (const TtyCanvasColor *)creader;
+      xleft = gx2 - gx1;
+      while (xleft-- != 0) {
+        TtyCanvasColor c = *reader++;
+        c.bgr.x = 0xff;
+        *plotter++ = c;
+      }
+      cplotter += xsfb;
+      creader += xs;
+    }
+  }
+}
+
+unrollloops void _TtyRgbxUpdate(struct Tty *tty) {
+  size_t gy1 = tty->updy1, gy2 = tty->updy2,
+         gx1 = tty->updx1, gx2 = tty->updx2,
+         xsfb = tty->xsfb, xs = tty->xs;
+  if (gy1 < gy2 && gx1 < gx2) {
+    size_t yleft = gy2 - gy1, xleft;
+    char *cplotter = tty->fb + gy1 * xsfb + gx1 * sizeof(TtyRgbxColor);
+    const char *creader = tty->canvas + gy1 * xs
+                                      + gx1 * sizeof(TtyCanvasColor);
+    TtyResetDirty(tty);
+    while (yleft-- != 0) {
+      TtyRgbxColor *plotter = (TtyRgbxColor *)cplotter;
+      const TtyCanvasColor *reader = (const TtyCanvasColor *)creader;
+      TtyCanvasColor ic;
+      TtyRgbxColor oc;
+      xleft = gx2 - gx1;
+      while (xleft-- != 0) {
+        ic.w = reader->w;
+        ++reader;
+        oc = (TtyRgbxColor){.rgb.r = ic.bgr.r, .rgb.g = ic.bgr.g,
+                            .rgb.b = ic.bgr.b, .rgb.x = 0xff};
+        plotter->w = oc.w;
+        ++plotter;
+      }
+      cplotter += xsfb;
+      creader += xs;
+    }
+  }
+}
+
+static void TtyGraphDrawBitmap(struct Tty *tty, size_t gy, size_t gx,
+                               TtyCanvasColor fg, TtyCanvasColor bg,
+                               const uint8_t *bitmap,
+                               size_t bm_ht, size_t bm_wid) {
+  size_t xs = tty->xs;
+  char *cplotter = tty->canvas + gy * xs
+                               + gx * sizeof(TtyCanvasColor);
+  size_t yleft = bm_ht, xleft;
+  while (yleft-- != 0) {
+    TtyCanvasColor *plotter = (TtyCanvasColor *)cplotter;
+    xleft = bm_wid;
+    while (xleft >= 8) {
+      uint8_t bits = *bitmap++;
+      *plotter++ = __builtin_add_overflow(bits, bits, &bits) ? fg : bg;
+      *plotter++ = __builtin_add_overflow(bits, bits, &bits) ? fg : bg;
+      *plotter++ = __builtin_add_overflow(bits, bits, &bits) ? fg : bg;
+      *plotter++ = __builtin_add_overflow(bits, bits, &bits) ? fg : bg;
+      *plotter++ = __builtin_add_overflow(bits, bits, &bits) ? fg : bg;
+      *plotter++ = __builtin_add_overflow(bits, bits, &bits) ? fg : bg;
+      *plotter++ = __builtin_add_overflow(bits, bits, &bits) ? fg : bg;
+      *plotter++ = __builtin_add_overflow(bits, bits, &bits) ? fg : bg;
+      xleft -= 8;
+    }
+    if (xleft) {
+      uint8_t bits = *bitmap++;
+      switch (xleft) {
+        default:
+          *plotter++ = __builtin_add_overflow(bits, bits, &bits) ? fg : bg;
```

Comment:
Cool fallthrough overflow pattern!

Reply:
I suspect that is about as optimized as one can go without bringing in SIMD instructions. :)